File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/controls_whileUntil.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import io.gopiper.piper.engine.Piper
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeControlsWhileUntil(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_controls_whileUntil")
        loopFlow = null
        val pointer = statement.getJSONObject("pointer")
        val loopBlock = pointer["loopBlock"].toString()
        val mode = pointer["mode"].toString()


        while (true) {
            val cond = try {
                executeStatement(pointer["cond"].toString()) as? Boolean ?: false
            } catch (e: Exception) {
                false
            }
            if (mode == "UNTIL") {
                if (cond) break
            } else {
                if (!cond) break
            }

            execute(loopBlock)
            if (isTerminated) break
            if (loopFlow == "BREAK") {
                loopFlow = null
                break
            } else if (loopFlow == "CONTINUE") {
                loopFlow = null
                continue
            }

            if (isFunctionReturn != null) break
        }
    } catch (e: Exception) {
        onError(e)
    }
}

File: piper/extensions/chrome/src/engine/statements/controls_whileUntil.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        this.loopFlow = null
        let pointer = statement.pointer
        let loopBlock = stringify(pointer.loopBlock)
        let mode = stringify(pointer.mode)

        while(true) {
            let cond
            try {
                cond = await this.executeStatement(stringify(pointer.cond))
            } catch (e) {
                cond = false
            }

            if (mode === "UNTIL") {
                if (cond) break
            } else {
                if (!cond) break
            }

            await this.execute(loopBlock)

            if (this.isTerminated) break
            if (this.loopFlow === "BREAK") {
                this.loopFlow = null
                break
            } else if (this.loopFlow === "CONTINUE") {
                this.loopFlow = null
                continue
            }

            if (this.isFunctionReturn !== null) break
        }
    } catch (e) {
        onError(e)
    }
}
